### DISCUSSION TITLE
Using _typeshed.Self for generic parameters

### DISCUSSION THREAD

**Discussion by Gobot1234** [python/typing#839] [Q&A]
When annotating a method using _typeshed.Self as the parameter to a Generic raises an error when returning and creating the object.

```python
from _typeshed import Self
from dataclasses import dataclass
from typing import Generic, Protocol, TypeVar

B = TypeVar("B", bound="Bar")


@dataclass
class Foo(Generic[B]):
	bar: B


class Bar:
	def baz(self: Self) -> Foo[Self]:  # error here
		return Foo(bar=self)  # another error here


reveal_type(Bar().baz())  # revealed type is Foo[Bar]
```

Mypy and pyright both fail to type check this but have the correct revealed types. I understand that the easy way to make this type check is to use ``(self: B) -> Foo[B]`` but I'm curious as to whether or not this is worth supporting in type checkers.

**Comment by srittau** [comment]
At the moment, `_typeshed.Self` is just a simple type var, a placeholder for if we get a proper `typing.Self`. As such, I don't think type checkers should handle it specially. But with a proper self type, I'd hope the above should work.
